apkid/rules/dex/obfuscators.yara

```diff
@@ -66,6 +67,46 @@ rule dexguard : obfuscator
     uint32(dex.header.data_offset + dex.header.data_size - 4) == 0
 }
 
+rule dexguard_a : obfuscator
+{
+  meta:
+    description = "DexGuard"
+    url         = "https://www.guardsquare.com/en/products/dexguard"
+    sample      = "41a9b44af8931d63812b4a23395b29279d2e055f357222dabed7153d4aee6299"
+    author      = "Eduardo Novella"
+
+  strings:
+    // 1-byte size + size-bytes obfuscated class + 1-byte NULL terminator
+    $Loaux   = { 07 4C 6F 2F (41|61) (55|75) (58|78) 3B 00 }  // Lo/[Aa][Uu][Xx];
+    $Locon   = { 07 4C 6F 2F (43|63) (4F|6F) (4E|6E) 3B 00 }  // Lo/[Cc][Oo][Nn];
+    $Lolcase = { 05 4C 6F 2F ?? 3B 00 }                       // Lo/[a-z];
+    $Loucase = { 05 4C 6F 2F ?? 3B 00 }                       // Lo/[A-Z];
+    $Loif    = { 06 4C 6F 2F ?? (46|66) 3B 00 }               // Lo/[iI][fF];
+    $Loif1U  = { 08 4C 6F 2F ?? 24 (49|69) (46|66) 3B 00 }    // Lo/[A-Z]$[iI][fF];
+    $Loif2UL = { 09 4C 6F 2F ?? ?? 24 (49|69) (46|66) 3B 00 } // Lo/[a-zA-Z][a-zA-Z]$[iI][fF];
+    $Lo2c    = { 06 4C 6F 2F ?? ?? 3B 00 }                    // Lo/[a-zA-z][a-zA-z];
+    $Lo2odd  = { 05 4C 6F 2F ?? ?? 3B 00 }                    // Lo/odd;
+    $Lo3odd  = { 05 4C 6F 2F ?? ?? ?? 3B 00 }                 // Lo/odd;
+
+    // Artifacts
+    $lib_runtime = "libruntime.so"
+    $dexguard    = "DexGuard" nocase
+    $guardsquare = "guardsquare" nocase
+    $lib_x       = /lib[a-z]{1}\.so/
+    $export      = /Java\_o\_\_[0-9a-f]{5}\_[0-9a-z]/
+
+  condition:
+      (
+        ($lib_runtime or $dexguard or $guardsquare)
+        or
+        (($Loaux or $Locon))
```

Comment:
I've used this trick before and have seen multiple obfuscators use invalid windows file names like aux and con (and nul!). The leading `'Lo/` might help make it more specific, but relying solely on class paths seems weak. Is there anything that could be added here? The rest of the rules seem fine.

Reply:
The leading `Lo/` helps a bit to fingerprint it. For the other questions, no.

Reply:
This specific rule seems to generic. Having `Lo/` as a class path is common for larger apps that've been proguarded. Using `aux` and `con` is somewhat common as an anti-disassembly technique (i've used it myself and have seen it a few times).

The other conditions seem quite good, but why are there so many entirely different conditions? It implies the rule should be split up into multiple varieties (a, b, c, d, etc.)

Reply:
> Lo/ is ProGuard; You might be wrong because this behaviour is more like DexGuard, isnt it?

This sentence doesn't compile. It's not clear what you're disagreeing with. You admit it's proguard and then say it's dexguard. Let me be specific: If you proguard an app with 15+ packages in any package scope using the default configuration you'll get `Lo/` package paths. Multiple obfuscators use the `aux` and `con` tricks because they're invalid windows filenames. Reliance on two potentially common package paths is not strong enough to say "this will never false positive".

> Splitting; I do like in the way is now.

This doesn't "parse". I can't be sure what you're saying. It sounds like you're saying "No, I don't want to change it because I like how it is now." If so, this is not an argument. You may have good reasons for keeping this rule together, but I can't imagine them, and you haven't given any.

> (($Loaux or $Locon)); You are probably right but it worked pretty well for me.

This is not an argument. In my experience, it's always possible to make rules more specific without losing sensitivity.